tree/tree/src/TTreeCache.cxx

```diff
@@ -614,10 +630,288 @@ Int_t TTreeCache::DropBranch(const char *bname, Bool_t subbranches /*= kFALSE*/)
+
+   fLastMiss = -1;
+   fFirstMiss = -1;
+
+   if (!fMissCache) {
+      fMissCache.reset(new MissCache());
+   }
+   fMissCache->clear();
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// For the event currently being fetched into the miss cache, find the IO
+/// (offset / length tuple) to pull in the current basket for a given branch.
+///
+/// Returns:
+/// - IOPos describing the IO operation necessary for the basket on this branch
+/// - On failure, IOPos.length will be set to 0.
+TTreeCache::IOPos TTreeCache::FindBranchBasket(TBranch &b) {
+   if (R__unlikely(b.GetDirectory() == 0)) {
+      //printf("Branch at %p has no valid directory.\n", &b);
+      return IOPos{0, 0};
+   }
+   if (R__unlikely(b.GetDirectory()->GetFile() != fFile)) {
+      //printf("Branch at %p is in wrong file (branch file %p, my file %p).\n", &b, b.GetDirectory()->GetFile(), fFile);
+      return IOPos{0, 0};
+   }
+
+   //printf("Trying to find a basket for branch %p\n", &b);
+   // Pull in metadata about branch; make sure it is valid
+   Int_t *lbaskets   = b.GetBasketBytes();
+   Long64_t *entries = b.GetBasketEntry();
+   if (R__unlikely(!lbaskets || !entries)) {
+      //printf("No baskets or entries.\n");
+      return IOPos{0, 0};
+   }
+   //Int_t blistsize = b.GetListOfBaskets()->GetSize();
+   Int_t blistsize = b.GetWriteBasket();
+   if (R__unlikely(blistsize <= 0)) {
+      //printf("Basket list is size 0.\n");
+      return IOPos{0, 0};
+   }
+
+   // Search for the basket that contains the event of interest.  Unlike the primary cache, we
+   // are only interested in a single basket per branch - we don't try to fill the cache.
+   Long64_t basketOffset = TMath::BinarySearch(blistsize, entries, fTree->GetReadEntry());
+   if (basketOffset < 0) { // No entry found.
+      //printf("No entry offset found for entry %ld\n", fTree->GetReadEntry());
+      return IOPos{0, 0};
+   }
+
+   // Check to see if there's already a copy of this basket in memory.  If so, don't fetch it
+   if ((basketOffset < blistsize) &&
+           b.GetListOfBaskets()->UncheckedAt(basketOffset)) {
+
+       //printf("Basket is already in memory.\n");
+       return IOPos{0, 0};
+   }
+
+   Long64_t pos = b.GetBasketSeek(basketOffset);
+   Int_t len = lbaskets[basketOffset];
+   if (R__unlikely(pos <= 0 || len <= 0)) {
+      /*printf("Basket returned was invalid (basketOffset=%ld, pos=%ld, len=%d).\n", basketOffset, pos, len);
+      for (int idx=0; idx<blistsize; idx++) {
+         printf("Basket entry %d, first event %d, pos %ld\n", idx, entries[idx], b.GetBasketSeek(idx));
+      }*/
+      return IOPos{0, 0};
+    } // Sanity check
+   // Do not cache a basket if it is bigger than the cache size!
+   if (R__unlikely(len > fBufferSizeMin)) {
+      //printf("Basket size is greater than the cache size.\n");
+      return IOPos{0, 0};
+   }
+
+   return IOPos{static_cast<ULong64_t>(pos), static_cast<UInt_t>(len)};
```

Comment:
Why elect to have the member of IOPos being unsigned rather than being the 'natural' type (and thus avoid the static cast while not losing anything since the interface takes/uses the 'natural' type)?